### PR TITLE
feat(global): apiCache implementation

### PR DESCRIPTION
**Pull request by alexfrst**
This PR implements the MVP for the apicache service:
- Get redis cache associated with a request
- Set redis cache associated with a request

### CODE REVIEW

File: __tests__/unit/index.ts [hunk 1]
```diff
@@ -1,5 +1,119 @@
-describe('Example', () => {
-  it('should success', () => {
-    expect(true);
-  })
+import * as redis from 'redis';
+import { Request } from 'express';
+import { ApiCache, ERedisFlag } from '../../src';
+
+describe('cacheService.ts', () => {
```
Comment (rimiti) [inline __tests__/unit/index.ts]: cache service ?

File: __tests__/unit/index.ts [hunk 2]
```diff
@@ -1,5 +1,119 @@
...
+    });
+    redisClient = redis.createClient();
+    apiCache = new ApiCache(redisClient);
+  });
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe('#getProviderCache', () => {
```
Comment (rimiti) [inline __tests__/unit/index.ts]: getProviderCache?

File: __tests__/unit/index.ts [hunk 3]
```diff
@@ -1,5 +1,119 @@
...
+
+      const redisGetAsyncSpy = jest
+        .spyOn(apiCache, 'redisGetAsync')
+        .mockResolvedValue(
+          'G6UBABwHzrk8JjkKvE79lCJbe0asK9WbzSc3QNVvNItOV9/7v/nCMbHA7O43ps4BvKZRBGGZnjR1WOS09YUIh0TRWnqJcLhZzxAeeAqKvXvTgm8SaBqebQQRSUsgjNRVcAyhP4HQ8v8CG1FAdbk/jQdAlJG/BBit6lel6RpI0Osvb0RD7uRlDTgWU6tlcbnUe14Sg7UOvamb3X+Zn0kOXtctFGoeVBmhqoqmjzENrEM+EY73j6QsxEl1uCc/L8AgTstzoDCz5+OxLHMWcqejn9ZN7Mv0Q3tfyHAi4CSMIth9yBk=',
+        );
+
+      const res = await apiCache.getCache({ query: {}, method: 'GET', path: '/langage/SGML/infos' } as Request);
+
+      expect(redisGetAsyncSpy).toBeCalledTimes(1);
```
Comment (rimiti) [inline __tests__/unit/index.ts:56]: calledWith?

File: __tests__/unit/index.ts [hunk 4]
```diff
@@ -1,5 +1,119 @@
...
+      expect(redisGetAsyncSpy).toBeCalledWith('get__langage/sgml/infos__');
```
Comment (rimiti) [inline __tests__/unit/index.ts]: Prefer to use .toHaveBeenNthCalledWith() instead of toBeCalledWith

File: __tests__/unit/index.ts [hunk 5]
```diff
@@ -1,5 +1,119 @@
...
+      const redisGetAsyncSpy = jest.spyOn(apiCache, 'redisGetAsync').mockResolvedValue(null);
+
+      const res = await apiCache.getCache({
+        query: {},
+        method: 'GET',
+        path: '/langage/invalid/infos',
+      } as Request);
+
+      expect(redisGetAsyncSpy).toBeCalledTimes(1);
+      expect(redisGetAsyncSpy).toBeCalledWith('get__langage/invalid/infos__');
```
Comment (rimiti) [inline __tests__/unit/index.ts]: Prefer to use .toHaveBeenNthCalledWith() instead of toBeCalledWith

File: __tests__/unit/index.ts [hunk 6]
```diff
@@ -1,5 +1,119 @@
...
+      };
+
+      await apiCache.setCache(
+        { query: {}, method: 'GET', path: '/langage/XML/infos' } as Request,
+        data,
+        1000 * 60 * 60 * 24 * 30,
+      );
+
+      expect(redisSetAsyncSpy).toBeCalledTimes(1);
+      expect(redisSetAsyncSpy).nthCalledWith(
```
Comment (rimiti) [inline __tests__/unit/index.ts]: Prefer to use .toHaveBeenNthCalledWith() instead of toBeCalledWith

File: __tests__/unit/index.ts [hunk 7]
```diff
@@ -1,5 +1,119 @@
...
+        path: '/langage/invalid/infos',
+      } as Request);
+
+      expect(redisGetAsyncSpy).toBeCalledTimes(1);
+      expect(redisGetAsyncSpy).toBeCalledWith('get__langage/invalid/infos__');
+      expect(res).toStrictEqual(undefined);
+    });
+  });
+
+  describe('#setProviderCache', () => {
```
Comment (rimiti) [inline __tests__/unit/index.ts]: setProviderCache?

File: __tests__/unit/index.ts [hunk 8]
```diff
@@ -1,5 +1,119 @@
...
+              },
+            },
+          },
+        },
+      };
+
+      await apiCache.setCache(
+        { query: {}, method: 'GET', path: '/langage/XML/infos' } as Request,
+        data,
+        1000 * 60 * 60 * 24 * 30,
```
Comment (rimiti) [inline __tests__/unit/index.ts]: Put directly the value

File: src/apiCache.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+import { RedisClient } from 'redis';
+import { promisify } from 'util';
+import { Request } from 'express';
+import { toPairs } from 'lodash';
+import * as flatted from 'flatted';
+import { ERedisFlag, IApiCacheConfiguration } from './types';
+import { defaultConfiguration } from './config';
+
+/**
+ * ApiCache class.
```
Comment (rimiti) [inline src/apiCache.ts]: @class / @description

File: src/apiCache.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+ */
+export class ApiCache {
+  private readonly redis: RedisClient;
+  private config: IApiCacheConfiguration;
+  public redisSetAsync: (key: string, value: string, flag?: ERedisFlag, expirationInMS?: number) => Promise<unknown>;
+  public redisGetAsync: (key: string) => Promise<string | null>;
+  private readonly compressAsync: (buffer: Buffer) => Promise<Buffer>;
+  private readonly decompressAsync: (buffer: Buffer) => Promise<Buffer>;
```
Comment (rimiti) [inline src/apiCache.ts]: private with privates
public with publics

File: src/apiCache.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    this.redisSetAsync = promisify(this.redis.set).bind(this.redis);
+    this.redisGetAsync = promisify(this.redis.get).bind(this.redis);
+    this.compressAsync = promisify(brotliCompress);
+    this.decompressAsync = promisify(brotliDecompress);
+  }
+
+  /**
+   * @description Builds the redis key associated with the corresponding request.
+   * @param {Request} req Express request
+   * @private
```
Comment (rimiti) [inline src/apiCache.ts]: @returns ?

File: src/apiCache.ts [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+   */
+  private buildKey(req: Request): string {
+    const generatedQuery = toPairs(req.query)
+      .map((pair: [string, unknown]) => pair.join(''))
```
Comment (rimiti) [inline src/apiCache.ts]: string instead of unknown no?

File: src/apiCache.ts [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    this.compressAsync = promisify(brotliCompress);
+    this.decompressAsync = promisify(brotliDecompress);
+  }
+
+  /**
+   * @description Builds the redis key associated with the corresponding request.
+   * @param {Request} req Express request
+   * @private
+   */
+  private buildKey(req: Request): string {
```
Comment (rimiti) [inline src/apiCache.ts]: All private methods at the end of your class.

File: src/apiCache.ts [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+      .map((pair: [string, unknown]) => pair.join(''))
+      .join('');
+
+    return `${this.config.prefix}${req.method}__${req.path.slice(1)}__${generatedQuery}`.toLowerCase();
+  }
+
+  /**
+   * @description Retrieves data from cache.
+   * @param {Request} req Express request to build redis key
+   * @returns {Promise<any>} Cache data
```
Comment (rimiti) [inline src/apiCache.ts]: any ?

File: src/apiCache.ts [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+   */
+  async getCache(req: Request): Promise<unknown> {
+    const rawData: string = (await this.redisGetAsync(this.buildKey(req))) as string;
+
+    if (rawData) {
+      const buffer: Buffer = await this.decompressAsync(Buffer.from(rawData, 'base64'));
```
Comment (rimiti) [inline src/apiCache.ts]: \n before the return

File: src/apiCache.ts [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+      return flatted.parse(buffer.toString());
+    }
+
+    return;
+  }
+
+  /**
+   * @description Stores data into cache.
+   * @param {Request} req Express request associated with the data
+   * @param {any} data - Data to cache
```
Comment (rimiti) [inline src/apiCache.ts]: any?

File: src/apiCache.ts [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+  async setCache(req: Request, data: unknown, durationInMS: number = this.config.expirationInMS): Promise<string> {
+    const compressedData: Buffer = await this.compressAsync(Buffer.from(flatted.stringify(data)));
+
+    return (await this.redisSetAsync(
+      this.buildKey(req),
+      compressedData.toString('base64'),
+      ERedisFlag.EXPIRATION_IN_MS,
+      durationInMS,
+    )) as string;
+  }
```
Comment (rimiti) [inline src/apiCache.ts]:
```ts
    return this.redisSetAsync(
      this.buildKey(req),
      compressedData.toString('base64'),
      ERedisFlag.EXPIRATION_IN_MS,
      durationInMS,
    ) as string;
  }
```

File: src/apiCache.ts (hunk 12 above)
Comment (alexfrst) [reply src/apiCache.ts]: the compiler refuses the type [string, string] it expects [string,unknown]

File: src/apiCache.ts (hunk 17 above)
Comment (alexfrst) [reply src/apiCache.ts]: redisSetAsync returns a promise to do the casting we need to wait for the result.

Comment (alexfrst) [reply src/apiCache.ts]: or I can cast it like that, as you prefer (Initial version or this one)`return this.redisSetAsync(
      this.buildKey(req),
      compressedData.toString('base64'),
      ERedisFlag.EXPIRATION_IN_MS,
      durationInMS,
    ) as Promise<string>;`

**Review by rimiti — CHANGES_REQUESTED** [review]
Some changes to do. Well done, be brave 💪

**Comment by rimiti** [comment]
Can you update the README.md file to explain with example how to use the package?

**Review by rimiti — APPROVED** [review]

**Pull request merged**